python_modules/dagster-pipes/dagster_pipes/__init__.py

```diff
@@ -468,6 +475,49 @@ def load_context(self, params: PipesParams) -> Iterator[PipesContextData]:
 T_MessageChannel = TypeVar("T_MessageChannel", bound="PipesMessageWriterChannel")
 
 
+class ExcThread(threading.Thread):
+    def run(self, *args, **kwargs):
+        self.exceptions = queue.Queue()
+
+        try:
+            super().run(*args, **kwargs)
+        except:
+            self.exceptions.put(sys.exc_info())
+
+    def join(self, *args, **kwargs):
+        super().join(*args, **kwargs)
+
+        while not self.exceptions.empty():
+            exc_info = self.exceptions.get()
+            sys.stderr.write(exc_info[1].with_traceback(exc_info[2]))
+
+
+class PipesLogWriterChannel(ABC):
+    pass
+
+
+T_LogChannel = TypeVar("T_LogChannel", bound=PipesLogWriterChannel)
+
+
+class PipesLogWriter(ABC, Generic[T_LogChannel]):
+    @abstractmethod
+    @contextmanager
+    def open(self, params: PipesParams) -> Iterator[T_LogChannel]: ...
+
+    @final
+    def get_opened_payload(self) -> PipesOpenedData:
+        return {"extras": self.get_opened_extras()}
+
+    def get_opened_extras(self) -> PipesExtras:
+        return {}
+
+    @abstractmethod
+    def make_channel(
+        self, params: PipesParams, stream: Literal["stdout", "stderr"]
+    ) -> T_LogChannel:
+        pass
```

Comment:
Unclear why this needs to be part of the top-level interface. `PipesMessageWriter` does not have this method in the interface and it is not calle externally.